subcommands/keys/tuf_updates_rotate_offline_key.go

```diff
@@ -0,0 +1,207 @@
+	newCiRoot.Signatures = make([]tuf.Signature, 0)
+	removeUnusedTufKeys(newCiRoot)
+	newProdRoot := genProdTufRoot(newCiRoot)
+
+	if shouldSign {
+		signNewTufRoot(curCiRoot, newCiRoot, newProdRoot, newCreds)
+	}
+
+	fmt.Println("= Uploading new TUF root")
+	tmpFile := saveTempTufCreds(keysFile, newCreds)
+	err = api.TufRootUpdatesPut(factory, txid, newCiRoot, newProdRoot, nil)
+	handleTufRootUpdatesUpload(tmpFile, keysFile, err)
+}
+
+func doTufUpdatesRotateOfflineTargetsKey(cmd *cobra.Command) {
+	factory := viper.GetString("factory")
+	txid, _ := cmd.Flags().GetString("txid")
+	keyTypeStr, _ := cmd.Flags().GetString("key-type")
+	keyType := ParseTufKeyType(keyTypeStr)
+	keysFile, _ := cmd.Flags().GetString("keys")
+	targetsKeysFile, _ := cmd.Flags().GetString("targets-keys")
+	shouldSign, _ := cmd.Flags().GetBool("sign")
+
+	if targetsKeysFile == "" {
+		targetsKeysFile = keysFile
+	}
+	if targetsKeysFile == "" {
+		subcommands.DieNotNil(errors.New(
+			"The --keys or --targets-keys option is required to rotate the offline TUF targets key.",
+		))
+	}
+	if shouldSign && keysFile == "" {
+		subcommands.DieNotNil(errors.New("The --keys option is required to sign the new TUF root."))
+	}
+
+	var (
+		creds, targetsCreds OfflineCreds
+		err                 error
+	)
+	if _, err := os.Stat(targetsKeysFile); err == nil {
+		targetsCreds, err = GetOfflineCreds(targetsKeysFile)
+		subcommands.DieNotNil(err)
+	} else if os.IsNotExist(err) {
+		targetsCreds = make(OfflineCreds, 0)
+	} else {
+		subcommands.DieNotNil(err)
+	}
+	subcommands.AssertWritable(targetsKeysFile)
+
+	if shouldSign {
+		if keysFile == targetsKeysFile {
+			creds = targetsCreds
+		} else {
+			creds, err = GetOfflineCreds(keysFile)
+			subcommands.DieNotNil(err)
+		}
+	}
+
+	updates, err := api.TufRootUpdatesGet(factory)
+	subcommands.DieNotNil(err)
+
+	curCiRoot, newCiRoot := checkTufRootUpdatesStatus(updates, true)
+
+	// Target "rotation" works like this:
+	// 1. Find the "online target key" - this the key used by CI, so we don't
+	//    want to lose it.
+	// 2. Generate a new key.
+	// 3. Set these keys in root.json.
+	// 4. Re-sign existing production targets.
+	onlineTargetsId, err := findOnlineTargetsId(factory, *newCiRoot)
+	subcommands.DieNotNil(err)
+	newKey, newCreds := replaceOfflineTargetsKey(newCiRoot, onlineTargetsId, targetsCreds, keyType)
+	fmt.Println("= New target keyid:", newKey.Id)
+	newCiRoot.Signatures = make([]tuf.Signature, 0)
+	removeUnusedTufKeys(newCiRoot)
+	newProdRoot := genProdTufRoot(newCiRoot)
+
+	fmt.Println("= Re-signing prod targets")
+	newTargetsSigs, err := resignProdTargets(factory, newCiRoot, onlineTargetsId, newCreds)
+	subcommands.DieNotNil(err)
+
+	if shouldSign {
+		signNewTufRoot(curCiRoot, newCiRoot, newProdRoot, creds)
+	}
+
+	fmt.Println("= Uploading new TUF root")
+	tmpFile := saveTempTufCreds(targetsKeysFile, newCreds)
+	err = api.TufRootUpdatesPut(factory, txid, newCiRoot, newProdRoot, newTargetsSigs)
```

Comment:
What happens if the `api.TufRootUpdatesPut` succeeds and then power cut happens so `handleTufRootUpdatesUpload` is never invoked? I guess a user should call `cancel` after that to be able to re-do the command from scratch.

Reply:
There are several options for the user:
- cancel and start anew.
- save the temporary file and proceed (we message about that).
- re-run the rotation command (a new key will be generated).

In general, this _transactional_ approach takes away the risk that the new key was applied on servers but not saved locally. Now, we don't apply until the user gives an explicit consent to do that; and we expect the user to make sure that the new key was properly stored before they apply TUF updates.

Reply:
makes sense.